### PR TITLE
[WIP] Issue #7: Roles and Permissions

### PR DESCRIPTION
**Pull request by lrowang**
Hi @lhridley ,

I've generated a WIP PR for #7. This is what I've done so far:

- Wrote the Behat tests for testing roles and permissions. As you can see from the commits, it took me a few tries to get it to work! ;) 
- Demonstrated that the Behat tests failed (see Pic1)
- I created the 2 new roles "staff" and "site administrator", but got a little hung up on the permissions part.

So, questions!

- There are A LOT of permissions. How do I know which permissions in the UI are encapsulated by "permission to author and publish content"? For example, I'm fairly certain things like `view published content`, `view own published content`, etc would be part of it. But what about things like `access the content over view page` and `administer news feeds`? (Technically a "feed" is just another form of content right?).
- On a related note: is there anyway to group all the permissions in the testing context so I can continue to use "author and publish content" as a <permission> argument in `roles-permissions.feature`? Or do we have to _list out every single permission_ in the scenario?
- Is there a more efficient way of selecting all/many permissions for a role? I did a little drush googling and saw that there is a way with `drush role-add-perm` + a list (but that still seemed sort of tedious. What about [permissions api](https://www.drupal.org/project/permissions_api), do we use that?

Phew okay that's all for now!

Pic1
issue7-behattest1

### CODE REVIEW

**Comment by lhridley** [comment]
Excellent first stab at the tests!

Since these will be custom steps, it is sort of within our control to determine how to structure the steps, and how to write the test code.  So let's talk about the most effective way to go about doing that.

The easiest way for us to set up the roles and permissions is through a Drupal Feature.  This will be a new concept for you, and one that you have not had exposure to before (at least I don't think you have).  Once you wrap your head around Features, you will love them -- they can be tricky.

Basically a Feature is a Drupal Module that you create through the user interface in a special way.  This module is dependent upon the Features module to run, and Features will need to be enabled for you to construct the feature.

Once you get the Feature created, you will simply enable the feature through a hook_update_n() (and also add it to the hook_install), and you will be good to go.

**Comment by lrowang** [comment]
Hi @lhridley - I watched a couple videos on Features on Drupalize.me, and here's my attempt at creating the Feature for saving the Roles and Permissions configurations. I realize think I might have accidentally enabled the feature via the UI (I was following along with the video) instead of using hook_update_n, but I can fix that on my next commit!

**Comment by lrowang** [comment]
Hi @lhridley - I enabled my new `roles_permissions` feature module in code! I ran `make clean-data` and `make install` to check and it looks like it's working!

File: www/sites/all/modules/custom/acs_master/acs_master.install [hunk 1]
```diff
@@ -119,6 +121,7 @@ function acs_master_uninstall() {
   module_disable($arr_reverse);
 }
 
+
```
Comment (lhridley) [inline www/sites/all/modules/custom/acs_master/acs_master.install]: You have an extra blank line that is causing the coding standards module (PHPCS) to flag this as a violation.  Please remove the blank line and recommit.

Also, I created the testing suite on another PR, you should be able to merge it into develop and rebase against develop, and you'll have a complete test suite.  We can discuss the structure of the tests and how you use the new context that I added for the Roles and Permissions.

**Comment by lrowang** [comment]
Hi @lhridley,

New commits! I think I resolved the rebase conflict that caused our Behat test files to be erroneously combined. Then I updated the permissions to make sure they matched what you had outlined in the Behat test, saved the config, updated the feature (with drush!), then ran the Behat tests! And hooray -- the tests passed locally!

This PR is now ready for your review!

**Comment by lhridley** [comment]
I tested both your hook updates and your hook install, and everything passes.  Looks Good!

One note, make sure you change the title of your PR to remove the "[WIP]" when it's ready for a final review, so that I know I can merge the PR and close the issue.  Otherwise we at Savas will review and give you feedback, but not merge and close anything.
